File: packages/swift-markdown-core/Sources/MarkdownCore/Walker/TreeDumper.swift
/// Produces the canonical diagnostic tree for immutable Markdown markup.
public enum TreeDumper {
    private struct Frame {
        var remainingChildren: Int
    }

    /// Returns the canonical diagnostic dump for `root` and its descendants.
    public static func dump(_ root: some Markup) -> String {
        var visitor = DumpVisitor()
        var frames: [Frame] = []
        var lines: [String] = []

        Walker().walk(root) { event, node in
            switch event {
            case .entering:
                let record = node.accept(&visitor)
                if frames.isEmpty {
                    lines.append(record.line)
                } else {
                    let parent = frames.count - 1
                    let prefix = frames.dropLast().map {
                        $0.remainingChildren > 0 ? "│   " : "    "
                    }.joined()
                    let connector = frames[parent].remainingChildren == 1 ? "└── " : "├── "
                    lines.append(prefix + connector + record.line)
                    frames[parent].remainingChildren -= 1
                }
                frames.append(Frame(remainingChildren: record.children))
            case .exiting:
                precondition(frames.last?.remainingChildren == 0)
                frames.removeLast()
            }
        }
        return lines.joined(separator: "\n") + "\n"
    }
}

extension Markup {
    /// Returns the canonical diagnostic dump for this markup subtree.
    public func dump() -> String { TreeDumper.dump(self) }
}

private struct DumpRecord {
    let line: String
    let children: Int
}

private struct DumpVisitor: MarkupVisitor {
    mutating func visit(_ node: DocumentRoot) -> DumpRecord {
        record("Document", node, children: node.content.count)
    }

    mutating func visit(_ node: BlockQuote) -> DumpRecord {
        record("BlockQuote", node, children: node.content.count)
    }

    mutating func visit(_ node: Paragraph) -> DumpRecord {
        record("Paragraph", node, children: node.content.count)
    }

    mutating func visit(_ node: Heading) -> DumpRecord {
        record("Heading", node, fields: ["level=\(node.level)"], children: node.content.count)
    }

    mutating func visit(_ node: ThematicBreak) -> DumpRecord {
        record("ThematicBreak", node)
    }

    mutating func visit(_ node: MarkdownCore.List) -> DumpRecord {
        record(
            "List",
            node,
            fields: [
                "flavor=\(node.flavor.rawValue)",
                "start=\(node.start.map(String.init) ?? "null")",
                "tight=\(boolean(node.tight))",
            ],
            children: node.items.count
        )
    }

    mutating func visit(_ node: ListItem) -> DumpRecord {
        record(
            "ListItem",
            node,
            fields: ["checked=\(node.checked.map(boolean) ?? "null")"],
            children: node.content.count
        )
    }

    mutating func visit(_ node: CodeBlock) -> DumpRecord {
        record(
            "CodeBlock",
            node,
            fields: [
                "info=\(optionalString(node.info))",
                "language=\(optionalString(node.language))",
                "literal=\(jsonString(node.literal))",
                "fenced=\(boolean(node.fenced))",
                "closed=\(boolean(node.closed))",
            ]
        )
    }

    mutating func visit(_ node: HTMLBlock) -> DumpRecord {
        record("HTMLBlock", node, fields: ["literal=\(jsonString(node.literal))"])
    }

    mutating func visit(_ node: FormulaBlock) -> DumpRecord {
        record("FormulaBlock", node, fields: ["literal=\(jsonString(node.literal))"])
    }

    mutating func visit(_ node: Table) -> DumpRecord {
        let alignments = node.alignments.map(\.rawValue).joined(separator: ",")
        return record(
            "Table",
            node,
            fields: ["alignments=[\(alignments)]"],
            children: 1 + node.rows.count
        )
    }

    mutating func visit(_ node: DirectiveBlock) -> DumpRecord {
        record(
            "DirectiveBlock",
            node,
            fields: directiveFields(node.name, node.attributes),
            children: (node.label == nil ? 0 : 1) + node.content.count
        )
    }

    mutating func visit(_ node: DirectiveLabel) -> DumpRecord {
        record("DirectiveLabel", node, children: node.content.count)
    }

    mutating func visit(_ node: FootnoteDefinition) -> DumpRecord {
        record(
            "FootnoteDefinition",
            node,
            fields: association(node.label, node.identifier),
            children: node.content.count
        )
    }

    mutating func visit(_ node: ReferenceDefinition) -> DumpRecord {
        record(
            "ReferenceDefinition",
            node,
            fields: association(node.label, node.identifier) + [
                "destination=\(jsonString(node.destination))",
                "title=\(optionalString(node.title))",
            ]
        )
    }

    mutating func visit(_ node: LinkReference) -> DumpRecord {
        record(
            "LinkReference",
            node,
            fields: association(node.label, node.identifier) + ["form=\(formName(node.form))"],
            children: node.content.count
        )
    }

    mutating func visit(_ node: ImageReference) -> DumpRecord {
        record(
            "ImageReference",
            node,
            fields: association(node.label, node.identifier) + ["form=\(formName(node.form))"],
            children: node.content.count
        )
    }

    /// The two fields five kinds carry identically, in contract order.
    private func association(_ label: String, _ identifier: String) -> [String] {
        ["label=\(jsonString(label))", "identifier=\(jsonString(identifier))"]
    }

    private func formName(_ form: ReferenceForm) -> String {
        switch form {
        case .full: "full"
        case .collapsed: "collapsed"
        case .shortcut: "shortcut"
        }
    }

    mutating func visit(_ node: Text) -> DumpRecord {
        record("Text", node, fields: ["literal=\(jsonString(node.literal))"])
    }

    mutating func visit(_ node: SoftBreak) -> DumpRecord { record("SoftBreak", node) }

    mutating func visit(_ node: LineBreak) -> DumpRecord { record("LineBreak", node) }

    mutating func visit(_ node: Code) -> DumpRecord {
        record("Code", node, fields: ["literal=\(jsonString(node.literal))"])
    }

    mutating func visit(_ node: HTML) -> DumpRecord {
        record("HTML", node, fields: ["literal=\(jsonString(node.literal))"])
    }

    mutating func visit(_ node: Formula) -> DumpRecord {
        record(
            "Formula",
            node,
            fields: ["mode=\(node.mode.rawValue)", "literal=\(jsonString(node.literal))"]
        )
    }

    mutating func visit(_ node: Emphasis) -> DumpRecord {
        record("Emphasis", node, children: node.content.count)
    }

    mutating func visit(_ node: Strong) -> DumpRecord {
        record("Strong", node, children: node.content.count)
    }

    mutating func visit(_ node: Strikethrough) -> DumpRecord {
        record("Strikethrough", node, children: node.content.count)
    }

    mutating func visit(_ node: Link) -> DumpRecord {
        record(
            "Link",
            node,
            fields: [
                "destination=\(optionalString(node.destination))",
                "title=\(optionalString(node.title))",
            ],
            children: node.content.count
        )
    }

    mutating func visit(_ node: Image) -> DumpRecord {
        record(
            "Image",
            node,
            fields: ["source=\(optionalString(node.source))", "title=\(optionalString(node.title))"],
            children: node.content.count
        )
    }

    mutating func visit(_ node: Directive) -> DumpRecord {
        record(
            "Directive",
            node,
            fields: directiveFields(node.name, node.attributes),
            children: node.label == nil ? 0 : 1
        )
    }

    mutating func visit(_ node: FootnoteReference) -> DumpRecord {
        record("FootnoteReference", node, fields: association(node.label, node.identifier))
    }

    mutating func visit(_ node: TableRow) -> DumpRecord {
        record(
            "TableRow",
            node,
            fields: ["isHeader=\(boolean(node.isHeader))"],
            children: node.cells.count
        )
    }

    mutating func visit(_ node: TableCell) -> DumpRecord {
        record("TableCell", node, children: node.content.count)
    }

    private func record(
        _ kind: String,
        _ node: any Markup,
        fields: [String] = [],
        children: Int = 0
    ) -> DumpRecord {
        let fieldText = fields.isEmpty ? "" : " " + fields.joined(separator: " ")
        return DumpRecord(
            line: "\(kind) \(scope(node.scope))\(fieldText) children=\(children)",
            children: children
        )
    }

    private func directiveFields(_ name: String, _ attributes: [DirectiveAttribute]?) -> [String] {
        guard let attributes else {
            return ["name=\(jsonString(name))", "attributes=null"]
        }
        let pairs = attributes.map { "\($0.name)=\(jsonString($0.value))" }.joined(separator: " ")
        return ["name=\(jsonString(name))", "attributes=[\(pairs)]"]
    }
}

private func scope(_ value: Scope) -> String {
    "scope=\(value.start.line):\(value.start.column)..\(value.end.line):\(value.end.column)"
}

private func boolean(_ value: Bool) -> String { value ? "true" : "false" }

private func optionalString(_ value: String?) -> String {
    value.map(jsonString) ?? "null"
}

private func jsonString(_ value: String) -> String {
    let hex = Array("0123456789abcdef")
    var result = "\""
    for scalar in value.unicodeScalars {
        switch scalar.value {
        case 0x22: result += "\\\""
        case 0x5c: result += "\\\\"
        case 0x08: result += "\\b"
        case 0x0c: result += "\\f"
        case 0x0a: result += "\\n"
        case 0x0d: result += "\\r"
        case 0x09: result += "\\t"
        case 0..<0x20:
            result += "\\u00\(hex[Int(scalar.value >> 4)])\(hex[Int(scalar.value & 0xf)])"
        default: result.unicodeScalars.append(scalar)
        }
    }
    return result + "\""
}

File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/walker/TreeDumper.kt
package com.nouprax.markdown.core

/** Produces the canonical diagnostic tree for immutable Markdown markup. */
public object TreeDumper {
    /** Returns the canonical diagnostic dump for [document] and its
     * descendants, resolving absolute scopes through the snapshot. */
    public fun dump(document: Document): String {
        val visitor = DumpVisitor()
        val remainingChildren = mutableListOf<Int>()
        val lines = mutableListOf<String>()

        Walker.walk(document) { event, node, scope ->
            when (event) {
                WalkEvent.ENTERING -> {
                    val record = node.accept(visitor)
                    val line = record.line(scope)
                    if (remainingChildren.isEmpty()) {
                        lines += line
                    } else {
                        val parent = remainingChildren.lastIndex
                        val prefix =
                            remainingChildren
                                .dropLast(1)
                                .joinToString(separator = "") { if (it > 0) "│   " else "    " }
                        val connector = if (remainingChildren[parent] == 1) "└── " else "├── "
                        lines += prefix + connector + line
                        remainingChildren[parent] -= 1
                    }
                    remainingChildren += record.children
                }

                WalkEvent.EXITING -> {
                    check(remainingChildren.removeAt(remainingChildren.lastIndex) == 0)
                }
            }
        }
        return lines.joinToString(separator = "\n", postfix = "\n")
    }
}

private data class DumpRecord(
    val kind: String,
    val fields: kotlin.collections.List<String>,
    val children: Int,
) {
    fun line(scope: Scope): String {
        val fieldText = if (fields.isEmpty()) "" else " ${fields.joinToString(" ")}"
        return "$kind ${scopeText(scope)}$fieldText children=$children"
    }
}

private class DumpVisitor : Visitor<DumpRecord> {
    override fun visitDocument(node: Document): DumpRecord = record("Document", children = node.content.size)

    override fun visitBlockQuote(node: BlockQuote): DumpRecord = record("BlockQuote", children = node.content.size)

    override fun visitParagraph(node: Paragraph): DumpRecord = record("Paragraph", children = node.content.size)

    override fun visitHeading(node: Heading): DumpRecord =
        record("Heading", fields = listOf("level=${node.level}"), children = node.content.size)

    override fun visitThematicBreak(node: ThematicBreak): DumpRecord = record("ThematicBreak")

    override fun visitList(node: List): DumpRecord =
        record(
            "List",
            fields =
                listOf(
                    "flavor=${node.flavor.token()}",
                    "start=${node.start ?: "null"}",
                    "tight=${node.tight}",
                ),
            children = node.items.size,
        )

    override fun visitListItem(node: ListItem): DumpRecord =
        record(
            "ListItem",
            fields = listOf("checked=${node.checked ?: "null"}"),
            children = node.content.size,
        )

    override fun visitCodeBlock(node: CodeBlock): DumpRecord =
        record(
            "CodeBlock",
            fields =
                listOf(
                    "mode=${node.mode.token()}",
                    "info=${optionalString(node.info)}",
                    "language=${optionalString(node.language)}",
                    "literal=${jsonString(node.literal)}",
                    "fenced=${node.fenced}",
                    "closed=${node.closed}",
                ),
        )

    override fun visitHTMLBlock(node: HTMLBlock): DumpRecord =
        record("HTMLBlock", fields = listOf("literal=${jsonString(node.literal)}"))

    override fun visitFormulaBlock(node: FormulaBlock): DumpRecord =
        record(
            "FormulaBlock",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visitTable(node: Table): DumpRecord =
        record(
            "Table",
            fields = listOf("alignments=[${node.alignments.joinToString(",") { it.token() }}]"),
            children = 1 + node.rows.size,
        )

    override fun visitTableRow(node: TableRow): DumpRecord =
        record(
            "TableRow",
            fields = listOf("isHeader=${node.isHeader}"),
            children = node.cells.size,
        )

    override fun visitTableCell(node: TableCell): DumpRecord = record("TableCell", children = node.content.size)

    override fun visitDirectiveBlock(node: DirectiveBlock): DumpRecord =
        record(
            "DirectiveBlock",
            fields = directiveFields(node.mode, node.name, node.attributes, node.label?.size),
            children = node.label.orEmpty().size + node.content.size,
        )

    override fun visitFootnoteDefinition(node: FootnoteDefinition): DumpRecord =
        record(
            "FootnoteDefinition",
            fields = listOf("id=${jsonString(node.label)}"),
            children = node.content.size,
        )

    override fun visitText(node: Text): DumpRecord =
        record("Text", fields = listOf("literal=${jsonString(node.literal)}"))

    override fun visitSoftBreak(node: SoftBreak): DumpRecord = record("SoftBreak")

    override fun visitLineBreak(node: LineBreak): DumpRecord = record("LineBreak")

    override fun visitCode(node: Code): DumpRecord =
        record(
            "Code",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visitHTML(node: HTML): DumpRecord =
        record("HTML", fields = listOf("literal=${jsonString(node.literal)}"))

    override fun visitFormula(node: Formula): DumpRecord =
        record(
            "Formula",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visitEmphasis(node: Emphasis): DumpRecord = record("Emphasis", children = node.content.size)

    override fun visitStrong(node: Strong): DumpRecord = record("Strong", children = node.content.size)

    override fun visitStrikethrough(node: Strikethrough): DumpRecord =
        record("Strikethrough", children = node.content.size)

    override fun visitLink(node: Link): DumpRecord =
        record(
            "Link",
            fields =
                listOf(
                    "destination=${optionalString(node.destination)}",
                    "title=${optionalString(node.title)}",
                ),
            children = node.content.size,
        )

    override fun visitImage(node: Image): DumpRecord =
        record(
            "Image",
            fields =
                listOf(
                    "source=${optionalString(node.source)}",
                    "title=${optionalString(node.title)}",
                ),
            children = node.content.size,
        )

    override fun visitDirective(node: Directive): DumpRecord =
        record(
            "Directive",
            fields = directiveFields(node.mode, node.name, node.attributes, node.label?.size),
            children = node.label.orEmpty().size,
        )

    override fun visitFootnoteReference(node: FootnoteReference): DumpRecord =
        record("FootnoteReference", fields = listOf("id=${jsonString(node.label)}"))
}

private fun record(
    kind: String,
    fields: kotlin.collections.List<String> = emptyList(),
    children: Int = 0,
): DumpRecord = DumpRecord(kind, fields, children)

private fun directiveFields(
    mode: PlacementMode,
    name: String,
    attributes: String?,
    labelCount: Int?,
): kotlin.collections.List<String> =
    listOf(
        "mode=${mode.token()}",
        "name=${jsonString(name)}",
        "attributes=${optionalString(attributes)}",
        "label=${labelCount ?: "null"}",
    )

private fun scopeText(value: Scope): String =
    "scope=${value.start.line}:${value.start.column}..${value.end.line}:${value.end.column}"

private fun optionalString(value: String?): String = value?.let(::jsonString) ?: "null"

private fun PlacementMode.token(): String = name.lowercase()

private fun ListFlavor.token(): String = name.lowercase()

private fun TableAlignment.token(): String = name.lowercase()

private fun jsonString(value: String): String =
    buildString {
        append('"')
        value.forEach { character ->
            when (character) {
                '"' -> {
                    append("\\\"")
                }

                '\\' -> {
                    append("\\\\")
                }

                '\b' -> {
                    append("\\b")
                }

                '\u000c' -> {
                    append("\\f")
                }

                '\n' -> {
                    append("\\n")
                }

                '\r' -> {
                    append("\\r")
                }

                '\t' -> {
                    append("\\t")
                }

                else -> {
                    if (character.code < 0x20) {
                        append("\\u")
                        append(character.code.toString(16).padStart(4, '0'))
                    } else {
                        append(character)
                    }
                }
            }
        }
        append('"')
    }
